pkg/cmd/get.go

```diff
@@ -0,0 +1,182 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/rest"
+)
+
+type GetOptions struct {
+	AllNamespaces  bool
+	IsCoreResource bool
+	APIPath        string
+	Resource       string
+	Token          string
+
+	KubeArchiveHost string
+
+	RESTConfig *rest.Config
+	kubeFlags  *genericclioptions.ConfigFlags
+}
+
+func NewGetOptions() *GetOptions {
+	return &GetOptions{
+		kubeFlags:       genericclioptions.NewConfigFlags(true),
+		KubeArchiveHost: "https://localhost:8081",
+	}
+}
+
+func NewGetCmd() *cobra.Command {
+	o := NewGetOptions()
+
+	cmd := &cobra.Command{
+		Use:   "get [RESOURCES]",
+		Short: "Command to get resources from KubeArchive",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var err error
+			err = o.Validate(args)
+			if err != nil {
+				return err
+			}
+
+			err = o.Complete(args)
+			if err != nil {
+				return err
+			}
+
+			err = o.Run()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	o.kubeFlags.AddFlags(cmd.Flags())
+	cmd.Flags().BoolVarP(&o.AllNamespaces, "all-namespaces", "A", o.AllNamespaces, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
+	cmd.Flags().StringVar(&o.KubeArchiveHost, "kubearchive-host", o.KubeArchiveHost, fmt.Sprintf("Host where the KubeArchive API Server is listening. Defaults to '%s'", o.KubeArchiveHost))
+
+	return cmd
+}
+
+func (o *GetOptions) Validate(args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("You must specify the type of resource to get.")
+	}
+
+	if len(args) > 1 {
+		return fmt.Errorf("You must specify just one resource to get.")
+	}
+
+	return nil
+}
+
+func (o *GetOptions) Complete(args []string) error {
+	o.Resource = args[0]
+	if strings.HasPrefix(o.Resource, "v1") {
+		o.IsCoreResource = true
+	}
+
+	o.APIPath = fmt.Sprintf("/apis/%s", o.Resource)
+	if o.IsCoreResource {
+		o.APIPath = fmt.Sprintf("/api/%s", o.Resource)
+	}
+
+	config, err := o.kubeFlags.ToRESTConfig()
+	if err != nil {
+		return err
+	}
+
+	o.RESTConfig = config
+
+	return nil
+}
+
+// TODO: when our API returns a List resource, we can remove the decode function
+func (o *GetOptions) getResources(host string, decode func([]byte) ([]unstructured.Unstructured, error)) ([]unstructured.Unstructured, error) {
+	url := fmt.Sprintf("%s%s", host, o.APIPath)
+
+	client, err := rest.HTTPClientFor(o.RESTConfig)
+	if err != nil {
+		return nil, err
+	}
+	response, err := client.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		bBytes, err := io.ReadAll(response.Body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("Status code != 200, but %d with body %s", response.StatusCode, string(bBytes))
```

Comment:
Does the `client.Get` call log information about the call? If not, this error message is not very useful. It says "status code != 200", but it doesn't say what you were doing. It doesn't even say that it was the return code from an HTTP call! It should at least say you were making an HTTP call to get a resource.

```suggestion
		return nil, fmt.Errorf("Return code from HTTP call to get resource %s was not 200, but %d with body %s", url, response.StatusCode, string(bBytes))
```

Reply:
I improved all the error handling, included this. I included something similar where the url, the status code and the string body is printed.